File: daily_problems/2024/07/0731/personal_submission/cf1041d_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
using namespace std;

signed main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    int n, h;
    cin >> n >> h;

    vector<int> ls(n), rs(n);
    for (int i = 0; i < n; i ++)
        cin >> ls[i] >> rs[i];
    
    int l = 0, r = 0, cur = h, res = 0, ans = 0;

    while (l < n) {
        while (r < n - 1 && ls[r+1] - rs[r] < cur) {
            res += rs[r] - ls[r];
            cur -= ls[r+1] - rs[r];
            r ++;
        }
        ans = max(ans, res + rs[r] - ls[r] + h);
        res -= rs[l] - ls[l];
        l ++;
        if (l < n)
            cur += ls[l] - rs[l-1];
    }
    cout << ans;

    return 0;
}

File: daily_problems/2024/07/0731/personal_submission/cf1041d_yawn_sean.py
# Submission link: https://codeforces.com/contest/1041/submission/273637870
def main():
    n, h = MII()
    ls = []
    rs = []
    for _ in range(n):
        l, r = MII()
        ls.append(l)
        rs.append(r)

    l, r = 0, 0
    cur = h

    res = 0
    ans = 0

    while l < n:
        while r < n - 1 and ls[r+1] - rs[r] < cur:
            res += rs[r] - ls[r]
            cur -= ls[r+1] - rs[r]
            r += 1
        ans = max(ans, res + rs[r] - ls[r] + h)
        res -= rs[l] - ls[l]
        l += 1
        if l < n:
            cur += ls[l] - rs[l-1]

    print(ans)